dev/com.ibm.ws.threading/src/com/ibm/ws/threading/CompletionStageFactory.java

```diff
@@ -0,0 +1,201 @@
+        else {
+            LibertyCompletableFuture<Void> future = new LibertyCompletableFuture<Void>(executor);
+            executor.execute(new RunAsync(action, future));
+            return future;
+        }
+    }
+
+    private static class RunAsync implements Runnable {
+        private final Runnable action;
+        private final LibertyCompletableFuture<Void> future;
+
+        RunAsync(Runnable action, LibertyCompletableFuture<Void> future) {
+            this.action = action;
+            this.future = future;
+        }
+
+        @FFDCIgnore(Throwable.class) // exception is from customer code path, not Liberty
+        @Override
+        public void run() {
+            try {
+                action.run();
+                future.complete(null);
+            } catch (Throwable x) {
+                future.completeExceptionally(x);
+            }
+        }
+    }
+
+    /**
+     * Creates a CompletableFuture which is backed (or partly backed if Java 8) by the Liberty global thread pool.
+     *
+     * @param <T>      type of result that is returned by the supplier.
+     * @param supplier operation that supplies a value.
+     * @return CompletableFuture that is backed (or partly backed if Java 8) by the Liberty global thread pool.
+     */
+    @Trivial
+    public <T> CompletableFuture<T> supplyAsync(Supplier<T> supplier) {
+        return supplyAsync(supplier, globalExecutor);
+    }
+
+    /**
+     * Creates a CompletableFuture which is backed (or partly backed if Java 8) by the specified
+     * executor as its default asynchronous execution facility.
+     * If the com.ibm.ws.concurrent bundle is enabled and the executor is a managed executor,
+     * then thread context is cleared/captured/propagated/restored for the supplier
+     * and all dependent stage actions per the configuration of the managed executor.
+     *
+     * @param <T>      type of result that is returned by the supplier.
+     * @param supplier operation that supplies a value.
+     * @param executor an executor. Null indicates to use the Liberty global thread pool.
+     * @return CompletableFuture that is backed (or partly backed if Java 8) by the specified executor.
+     */
+    public <T> CompletableFuture<T> supplyAsync(Supplier<T> supplier, Executor executor) {
+        if (executor instanceof CompletionStageExecutor)
+            return ((CompletionStageExecutor) executor).supplyAsync(supplier);
+
+        executor = executor == null ? globalExecutor : executor;
+        if (JAVA8)
+            return CompletableFuture.supplyAsync(supplier, executor); // at least run the initial stage on the executor
+        else {
+            LibertyCompletableFuture<T> future = new LibertyCompletableFuture<T>(executor);
+            // TODO once Java 8 is dropped, replace reflection with:
+            // TODO future.completeAsync(supplier, executor)
```

Comment:
Would a multi-release jar work here? Usually, I think MR jars are more work (dual maintenance) than they're worth, but maybe this case might be more beneficial?

Reply:
Yes - good idea - that could be a way to further improve it, but I'll leave that sort thing up to the performance team for future consideration.